src/nvim/api/vim.c

```diff
@@ -164,6 +165,39 @@ Object vim_eval(String str, Error *err)
   return rv;
 }
 
+/// Call the given function with the given arguments stored in an array.
+///
+/// @param functionName The function to call
+/// @param args the functions arguments packed in an Array
+/// @param[out] err Details of an error that may have occurred
+/// @return The result of the function call if it was successful
+Object vim_function(String functionName, Array args, Error *err)
+  FUNC_ATTR_DEFERRED
+{
+  Object rv = OBJECT_INIT;
+  typval_T r;
+  size_t i;
+  typval_T argv[args.size];
+
+  try_start();
+  // Convert the arguments in args from Object to typval_T values
+  for (i = 0; i < args.size; i++) {
+    if (!object_to_vim(args.items[i], &argv[i], err)) {
+      break;
+    }
+  }
+
+  if (!err->set && func_call((char_u*) functionName.data, argv, NULL, &r) == FAIL) {
```

Comment:
Not sure, but messagepack strings may not be null terminated. That may be the cause of your segfault. It could also be the `argv` problem noted above. Also please declare `r` here.